Issue #3083877 by ribel, agami4: As a LU I want to have group and profile layout in new sky flavor

ribel opened the pull request:
## Problem
<describe the problem you're trying to solve>

## Solution
<describe the solution you're created>

## Issue tracker
https://www.drupal.org/project/social/issues/3083877

## How to test
- [ ] <enter multiple how to test steps here>

## Release notes
<describe the release notes>

## Change Record
If this Pull Request changes the way that developers should do things or introduces a new API for developers then a change record to document this is needed. Please provide a draft for a change record or a link to an unpublished change record below. [Existing change records](https://www.drupal.org/list-changes/social) can be consulted as example.

Kingdutch commented on modules/social_features/social_group/social_group.module:
```diff
@@ -317,10 +317,20 @@ function social_group_preprocess_group(array &$variables) {
+  }
+  if ($group->getGroupType()->hasContentPlugin('group_node:topic')) {
+    $variables['group_topics'] = count($group->getContent('group_node:topic'));
+  }
```
❗️ This is a serious performance regression. We're basically loading all the group content here before counting it. 

The code that was removed was more efficient. We should probably create a `GroupStatistics` service in the `social_group` module and create method for each type there. Then we can deprecate `SocialGroupMembersCount` and route the `SocialGroupMembersCount::getGroupMemberCount` call to `GroupStatistics::memberCount`.

rollki replied:
@Kingdutch 
I agree with you. I already fixed this performance issue [here](https://github.com/goalgorilla/open_social/pull/1515/commits/cd9ededd5a30a0f2fcfb176a63bc12b5972d5a7f).
I created a new service `GroupStatistics` which has 2 methods:
* `nodeCount` for an event and topic types
* `memberCount` for the `group_membership`
Also I removed old service `SocialGroupMembersCount` and used Dependency Injection in `SocialGroupListBuilder` for the `groupStatistics` service instead of `\Drupal->service()`

Kingdutch commented on modules/social_features/social_group/social_group.module:
```diff
@@ -317,10 +317,20 @@ function social_group_preprocess_group(array &$variables) {
+    $variables['group_topics'] = count($group->getContent('group_node:topic'));
+  }
+  if ($group->getGroupType()->id() !== 'closed_group' || $group->getMember($account) || $group->hasPermission('manage all groups', $account)) {
+    $variables['group_stream_url'] = Url::fromUserInput('/group/' . $group->id() . '/stream');
```
We shouldn't be using `fromUserInput` but instead use the routename with the proper argument. This ensures that we don't create issues if the URL is changed.

rollki replied:
@Kingdutch
Fixed for all cases in this module.

Kingdutch commented on modules/social_features/social_group/social_group.module:
```diff
@@ -1325,6 +1336,12 @@ function social_group_block_access(Block $block, $operation, AccountInterface $a
+    if (theme_get_setting('style') !== 'sky') {
+      return AccessResult::forbidden();
+    }
+  }
```
We control the block so we should probably just add this condition to the block.

Additionally, this creates a dependency of the block on the theme, which shouldn't happen. So we should find a way around this.

rollki replied:
@Kingdutch 
Added to the `blockAccess`.

Kingdutch commented on modules/social_features/social_landing_page/social_landing_page.module:
```diff
@@ -62,10 +62,16 @@ function social_landing_page_theme() {
   ];
+  $theme_templates['group__featured__sky'] = [
+    'base hook' => 'group',
+  ];
```
We're creating a new template/theme couple here. We probably want to adjust the template suggestions for the `group__featured` template instead.

ribel replied:
Templates were moved to socialblue theme (the same as for content).

Kingdutch commented on modules/social_features/social_landing_page/social_landing_page.module:
```diff
@@ -62,10 +62,16 @@ function social_landing_page_theme() {
   ];
+  $theme_templates['profile__featured__sky'] = [
+    'base hook' => 'profile',
+  ];
```
We're creating a new template/theme couple here. We probably want to adjust the template suggestions for the `profile__featured` template instead.

ribel replied:
This Template was moved to socialblue theme (the same as for content).

Kingdutch commented on modules/social_features/social_landing_page/templates/group--featured--sky.html.twig:
```diff
@@ -0,0 +1,65 @@
+{% extends 'group--featured.html.twig' %}
```
I feel like this file has a lot in common with changes we make in #1511 Perhaps we can deduplicate this somehow?

Additionally, we shouldn't have `--sky` things in modules since this is a theme thing.

ribel replied:
A template was refactored and simplified.

Kingdutch commented on modules/social_features/social_landing_page/templates/node--featured--sky.html.twig:
```diff
@@ -1,17 +1,18 @@
 {% extends "node--featured.html.twig" %}
```
I missed where this file was introduced, but I think we should keep `--sky` templates in the `socialblue` template because they're a specialcase of that theme and not part of this module.

ribel replied:
Fixed and removed from here.

Kingdutch commented on modules/social_features/social_landing_page/templates/profile--featured--sky.html.twig:
```diff
@@ -0,0 +1,21 @@
+{% extends 'profile--featured.html.twig' %}
```
I think we should keep `--sky` templates in the `socialblue` theme because they're a special case of that theme and not part of this module.

ribel replied:
Fixed.

Kingdutch commented on modules/social_features/social_profile/social_profile.module:
```diff
@@ -377,6 +379,18 @@ function social_profile_preprocess_profile(array &$variables) {
+    'uid' => $user->id(),
+    'type' => 'event',
+  ]));
+  $variables['profile_groups'] = count(\Drupal::service('social_group.helper_service')->getAllGroupsForUser($user->id()));
```
❗️  This is a performance red flag because we're loading a lot of content here to count it. We should instead just create an entity count query. Perhaps create a `UserStatistics` service much like `GroupStatistics`

rollki replied:
@Kingdutch 
Fixed. Created the UserStatistics service.

Kingdutch commented on modules/social_features/social_profile/social_profile.module:
```diff
@@ -660,24 +674,24 @@ function social_profile_social_user_account_header_account_links(array $context)
+      ] + Url::fromRoute('entity.profile.type.user_profile_form', [
+        'user' => $context['user']->id(),
+        'profile_type' => 'profile',
+      ])->toRenderArray(),
```
Did the coding standards change?

rollki replied:
Fixed.

Kingdutch commented on modules/social_features/social_profile/social_profile.module:
```diff
@@ -711,3 +725,17 @@ function social_profile_social_user_account_header_items_alter(array &$menu_item
+  }
+  // No opinion for other situations.
+  return AccessResult::neutral();
+}
```
We own the block, so we don't need an external access handler. Additionally, we should decouple this from the theme in some way.

Kingdutch commented on modules/social_features/social_profile/src/Plugin/Block/ProfileStatisticBlock.php:
```diff
@@ -0,0 +1,106 @@
+        ->getViewBuilder('profile')
+        ->view($profile, 'statistic');
+      $build['content']['#cache']['tags'] = $this->getCacheTags();
+      $build['content']['#cache']['contexts'] = $this->getCacheContexts();
```
I like all the code above! However, I don't think we need these tags and contexts because they duplicate the block lower in the tree so this has no real effect other than some bookkeeping overhead.

Kingdutch commented on modules/social_features/social_profile/src/Plugin/Block/ProfileStatisticBlock.php:
```diff
@@ -0,0 +1,106 @@
+    $profile = $storage->loadByUser($account, 'profile');
+
+    if ($profile) {
+      $build['content'] = $this->entityTypeManager
```
I think we can just set $build to equal the result of this and don't need to nest this in `content`.

Kingdutch commented on themes/socialbase/templates/layout/page.html.twig:
```diff
@@ -39,16 +39,20 @@
+  {% if not style == 'sky' %}
+    {% if page.breadcrumb %}
+      {{ page.breadcrumb }}
+    {% endif %}
```
Socialbase shouldn't be aware of the `sky` flavour. Perhaps we can remove the breadcrumbs through a preprocess hook in the `socialblue` theme. 

@xinyuma should also probably think about how we want to use breadcrumbs throughout all flavours, I think this is still something we're inconsistent with even though they can be useful.

Kingdutch commented on themes/socialbase/templates/layout/page.html.twig:
```diff
@@ -39,16 +39,20 @@
+  {% if not style == 'sky' %}
+    {% if page.secondary_navigation %}
+      {{ page.secondary_navigation }}
+    {% endif %}
```
Socialbase shouldn't be aware of the sky flavour

Kingdutch commented on themes/socialbase/templates/layout/page.html.twig:
```diff
@@ -63,14 +67,22 @@
         {{ page.title }}
       {% endif %}
 
+      {% if style == 'sky' %}
```
Socialbase shouldn't be aware of the sky flavour

Kingdutch commented on themes/socialbase/templates/layout/page.html.twig:
```diff
@@ -63,14 +67,22 @@
-          {% if page.complementary_bottom %}
-            {{ page.complementary_bottom }}
-          {% endif %}
+          <div class="complementary--content-merged card">
```
Does this `div` truly add value here? Can we not just style the `region--complementary` instead?

Kingdutch commented on themes/socialblue/components/03-molecules/navigation/navbar/navbar-secondary--sky.js:
```diff
@@ -0,0 +1,65 @@
+(function ($) {
+
+  Drupal.behaviors.navbarSecondaryScrollable = {
```
Please add a top level comment to this function to explain what it's used for so we still know this in 6 months time.

Additionally, please add comments throughout the function to make it easier for other people than the author to follow what's going on. Comments that document **why** something happens are more useful than **what** happens because the latter can be seen by reading the code.

agami4 replied:
The comment was added to the `navbar-secondary--sky.js`

Kingdutch commented on themes/socialblue/socialblue.info.yml:
```diff
@@ -10,6 +10,7 @@ core: 8.x
 libraries:
   - socialblue/brand
   - socialblue/site-footer
+  - socialblue/nav-secondary--sky
```
This means we're adding the library always, even when the sky flavour isn't used.

Kingdutch commented on themes/socialblue/src/Plugin/Preprocess/Page.php:
```diff
@@ -0,0 +1,45 @@
+
+namespace Drupal\socialblue\Plugin\Preprocess;
+
+use Drupal\socialbase\Plugin\Preprocess\Page as PageBase;
```
We should check if we should overwrite the Socialbase Page Preprocessor here because I think they are all executed by Bootstrap, this would case the SocialBase one to be executed twice. Use the Bootstrap PreprocessBase instead.

ribel replied:
In this case, we need to use `$variables['content_attributes']`, which was defined only in socialbase, but not exists in bootstrap. It gives error when I try to extend it directly from bootstrap, so we need to leave it like this.

Kingdutch commented on themes/socialblue/src/Plugin/Preprocess/Page.php:
```diff
@@ -0,0 +1,45 @@
+    parent::preprocess($variables, $hook, $info);
+
+    $style = theme_get_setting('style');
+    if ($style && $style === 'sky') {
```
The first check is unneccessary here I think. Any value that would return false there would also not `=== 'sky'`.

Kingdutch commented on themes/socialblue/templates/group/group--statistic--sky.html.twig:
```diff
@@ -0,0 +1,40 @@
+      <span class="card__counter-quantity">{{ group_members }}</span>
+      <span class="card__counter-text">{% trans %}member{% plural group_members %}members{% endtrans %}</span>
+    </li>
+  </ul>
```
Can we in some way make this more generic or fieldable? It think we're using this structure more often going forward and we could set this up in such a way that we don't need to adjust a template when we want to add a statistic (or have an extension add a statistic or change what's shown).

Kingdutch commented on themes/socialblue/templates/group/group--statistic--sky.html.twig:
```diff
@@ -0,0 +1,40 @@
+      {% endif %}
+    </div>
+  </div>
+{% endif %}
```
This feels like a lot of logic that probably shouldn't be in a template file but simply be created with a render array.

Kingdutch commented on themes/socialblue/templates/profile/profile--profile--statistic--sky.html.twig:
```diff
@@ -0,0 +1,51 @@
+{{ attach_library('socialbase/hero') }}
```
The same questions about making this more generic exist as for `themes/socialblue/templates/group/group--statistic--sky.html.twig`.

ribel commented:
Hi @Kingdutch, we fixed all the things after your initial review and fixed issues found by Xinyu.

The only thing which is left from your general comment to not have a condition about `sky` in modules are left in of Group and Profile Statistic block access. 
I've researched but looks like there are no real options in Drupal to avoid it. We can create custom ConditionPlugin, but this will not change too much since anyway this code will be located in the module. So I would propose to leave it like it is for now, and maybe later We can just display the same block in the default style as well.

Kingdutch commented on modules/social_features/social_group/src/Plugin/Block/GroupStatisticBlock.php:
```diff
@@ -0,0 +1,110 @@
+  public function build() {
+    $build = [];
+
+    $group = _social_group_get_current_group();
```
This should use the block context instead. The block context should probably be set to `required`. You may need to set the default configuration for this block to pick the group from the URL.

An example can be seen in https://bitbucket.org/goalgorilla/social_dashboard/src/master/src/Plugin/Block/DashboardNavigationBlock.php

ribel replied:
This was fixed in PR #1680

Kingdutch commented on modules/social_features/social_landing_page/config/install/core.entity_view_display.profile.profile.featured.yml:
```diff
@@ -16,7 +16,7 @@ dependencies:
     - field.field.profile.profile.field_profile_self_introduction
     - field.field.profile.profile.field_profile_show_email
-    - image.style.social_featured
+    - image.style.social_large
```
The `social_featured` image style was created for this view mode, so if we need to change it because we're changing how landing pages look then we should probably just change the image style itself.

Additionally this will also impact "old-style" landing pages.

ribel replied:
We need to change image style on featured teasers only for profiles (for both old and sky - this was agreed with Xinyu).
But we still use `social_featured` image style for featured content and groups.
So in this way, it makes the most sense to change image style to ones that fit with a new design.
Also, it shouldn't break anything, since CSS for both old and new styles were updated as well

Kingdutch commented on modules/social_features/social_profile/social_profile.module:
```diff
@@ -376,7 +376,15 @@ function social_profile_preprocess_profile(array &$variables) {
+  $user_statistics = \Drupal::service('social_profile.user_statistics');
+  $variables['profile_topics'] = $user_statistics->nodeCount($user->id(), 'topic');
+  $variables['profile_events'] = $user_statistics->nodeCount($user->id(), 'event');
+  $variables['profile_groups'] = count(\Drupal::service('social_group.helper_service')->getAllGroupsForUser($user->id()));
```
❗️  This now runs for all view modes and can have a negative effect on performance on user listings (where this data is not used).

ribel replied:
This was fixed in PR #1680

Kingdutch commented on themes/socialblue/color/color.inc:
```diff
@@ -5,6 +5,16 @@
+// Add brand css files for new Sky style.
+$theme_path = drupal_get_path('theme', 'socialblue');
+$sky_brand_css = 'assets/css/brand--sky.css';
+if (file_exists($theme_path . '/' . $sky_brand_css)) {
```
Isn't this always true?

ribel replied:
Good catch - I simplified it and fixed in PR #1680

Kingdutch commented on themes/socialblue/components/03-molecules/dropdown/dropdown--sky.scss:
```diff
@@ -0,0 +1,17 @@
+  .hero-footer__cta .open .caret,
+  .social-gtranslate-list--desktop.open > a:after {
+    transform: rotate(180deg);
+  }
```
I think this is a change we can make in both flavours.

Additionally the following can make it look at bit more friendly:
```
.dropdown .caret {
    transition: transform 0.1s;
}
```

agami4 replied:
@Kingdutch, I've moved these styles in the SocialBase theme and add `transition`.

ribel replied:
It was fixed in PR #1680

Kingdutch commented on themes/socialblue/components/04-organisms/comment/comment--sky.scss:
```diff
@@ -0,0 +1,62 @@
+@import "settings";
+
+.socialblue--sky {
+  [id='section-comments'] {
```
```suggestion
  #section-comments {
```

agami4 replied:
Fixed

ribel replied:
It was fixed in PR #1680

Kingdutch commented on themes/socialblue/package-lock.json:
```diff
@@ -0,0 +1,9873 @@
+{
```
Please remember that we use `yarn` and not `npm` for our theme package management.

agami4 replied:
The `package-lock.json` was removed

ribel replied:
It was fixed in PR #1680

Kingdutch commented on themes/socialblue/socialblue.theme:
```diff
@@ -32,5 +32,34 @@ function socialblue_library_info_alter(&$libraries, $extension) {
+
+    // Add extra JS for secondary navigation in Sky style.
+    $navbar_js = 'assets/js/navbar-secondary--sky.min.js';
+    if (file_exists($theme_path . '/' . $navbar_js)) {
```
Isn't this always true and shouldn't this depend on whether the Sky flavour is enabled?

ribel replied:
Good catch - I simplified it and fixed in PR #1680

ribel commented on modules/social_features/social_group/social_group.services.yml:
```diff
@@ -32,6 +32,6 @@ services:
     arguments: ['@config.factory']
-  social_group.group_members_count:
-    class: Drupal\social_group\SocialGroupMembersCount
+  social_group.group_statistics:
```
This was fixed in PR #1680

ribel commented on modules/social_features/social_group/src/GroupStatistics.php:
```diff
@@ -36,14 +36,44 @@ public function __construct(Connection $connection) {
+   *   The number of members.
    */
-  public function getGroupMemberCount(GroupInterface $group) {
+  public function memberCount(GroupInterface $group) {
```
This was fixed in PR #1680 - I reverted the old method name
